## Regression in Machine Learning Quiz Question 12-20
### By Adedamola Babayato
### ID: 157372abee81f000

In [1]:
# Import the Libraries
import pandas as pd
import matplotlib.pyplot  as plt
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import Ridge, Lasso, LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [2]:
# Read in the csv file
df = pd.read_csv('energydata_complete.csv')

In [3]:
# check the head info
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

In [5]:
# drop irrelevant features like the date column
df.drop("date", axis = 1, inplace = True)

In [6]:
# check the shape of the dataset
df.shape

(19735, 28)

### Regression

In [7]:
# firstly we normalize the dataset to common scale using the min max scaler
# Instantiate an object for the min max scaler

scaler = MinMaxScaler()

In [8]:
# Create a normalized dataframe

normalized_df = pd.DataFrame(scaler.fit_transform(df), columns = df.columns)

In [9]:
# create a separate object for the X and y variables

X = normalized_df[["T2"]]

y = normalized_df["T6"]

Scikit learn model_selection module has a train_test_split capability for separating the dataset into training and testing portion. We would reserve 30% of the data for testing by setting test size = 0.3

In [10]:
# Split the dataset into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

### Training the Model

We can train a linear regression model using our training data.

**Create an instance of a LinearRegression() model**

In [11]:
lin_mod= LinearRegression()

**Next, we train the linear regression model with training data by fitting it**

In [12]:
lin_mod.fit(X_train,y_train)

LinearRegression()

**Since linear regression produces coefficients of the linear equation that beat fit the data, we can go ahead to print out the coefficients of the model**

In [13]:
print('Coefficients: \n', lin_mod.coef_)

Coefficients: 
 [0.8910771]


#### Predicting Test Data
Now that we have fit our model, let's evaluate its performance by predicting the labels of test data with it

**We would use lin_mod.predict() to predict labels(Y) of the X_test set of the data.**

In [14]:
predictions = lin_mod.predict(X_test)

## Questions

### Question 12:
#### From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the R^2 value in two d.p.?

In [15]:
# Evaluate the performance of the model using R_squared
from sklearn.metrics import r2_score
R = r2_score(y_test, predictions)
print(round(R, 2))

0.64


The r2 value is 0.64

### Question 13:
#### Normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set. What is the Mean Absolute Error (in two decimal places)?

In [16]:
# Drop the lights column
normalized_df.drop("lights", axis = 1, inplace = True)

In [17]:
# create a separate object for the X and y variables
new_X = normalized_df.drop("Appliances", axis =1)
new_y = normalized_df["Appliances"]

In [18]:
# Print the column names
normalized_df.columns

Index(['Appliances', 'T1', 'RH_1', 'T2', 'RH_2', 'T3', 'RH_3', 'T4', 'RH_4',
       'T5', 'RH_5', 'T6', 'RH_6', 'T7', 'RH_7', 'T8', 'RH_8', 'T9', 'RH_9',
       'T_out', 'Press_mm_hg', 'RH_out', 'Windspeed', 'Visibility',
       'Tdewpoint', 'rv1', 'rv2'],
      dtype='object')

In [19]:
# Split the dataset into training and test set
X_train, X_test, y_train, y_test = train_test_split(new_X, new_y, test_size = 0.3, random_state = 42)

In [20]:
# Create an instance for the model
pred = LinearRegression()

In [21]:
# Train the model
pred.fit(X_train, y_train)

LinearRegression()

In [22]:
# make prediction with the model
prediction_val = pred.predict(X_test)

In [23]:
# Evaluate the performance of the model using Mean Absolute Error (MAE) metric
MAE = mean_absolute_error(y_test, prediction_val)
print(round(MAE, 2))

0.05


The MAE value is 0.05

### Question 14:
#### What is the Residual Sum of Squares (in two decimal places)?

In [24]:
# Calculate the Residual Sum of Squares
RSS = np.sum(np.square(y_test - prediction_val))
round(RSS, 2)

45.35

The residual sum of squares is 45.35

### Question 15:
#### What is the Root Mean Squared Error (in three decimal places)?

In [25]:
# Calculate the Root Mean Squared Error
RMSE = np.sqrt(mean_squared_error(y_test, prediction_val))
round(RMSE, 3)

0.088

The RMSE value is 0.088

### Question 16:
#### What is the Coefficient of Determination (in two decimal places)?

In [26]:
# Calculate the R_squared
R_squared = r2_score(y_test, prediction_val)
round(R_squared, 2)

0.15

The coefficient of determination is 0.15

### Question 17:
#### Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?

In [27]:
def get_weights_df(pred, feat, col_name):
    #this function returns the weight of every feature
    weights = pd.Series(pred.coef_, feat.columns).sort_values()
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ['Features', col_name]
    weights_df[col_name].round(3)
    return weights_df

# Execute the get_weight function and store in dataframe
linear_weights_df = get_weights_df(pred, X_train, 'Linear_weight')

linear_weights_df

Features  Linear_weight
0           rv2  -5.776231e+10
1          RH_2  -4.568235e-01
2         T_out  -3.218456e-01
3            T2  -2.362756e-01
4            T9  -1.899430e-01
5          RH_8  -1.576318e-01
6        RH_out  -7.763937e-02
7          RH_7  -4.457684e-02
8          RH_9  -3.980359e-02
9            T5  -1.566380e-02
10           T1  -3.281051e-03
11  Press_mm_hg   6.861092e-03
12           T7   1.030094e-02
13   Visibility   1.229842e-02
14         RH_5   1.600668e-02
15         RH_4   2.638746e-02
16           T4   2.901612e-02
17    Windspeed   2.920236e-02
18         RH_6   3.806075e-02
19         RH_3   9.606639e-02
20           T8   1.020090e-01
21    Tdewpoint   1.177704e-01
22           T6   2.364340e-01
23           T3   2.906878e-01
24         RH_1   5.535871e-01
25          rv1   5.776231e+10

The features with the lowest and highest weights are RH_2 and RH_1 respectively.

### Question 18:
#### Train a ridge regression model with an alpha value of 0.4. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [28]:
ridge_Reg = Ridge(alpha=0.04)
ridge_Reg.fit(X_train, y_train)
ridge_pred = ridge_Reg.predict(X_test)
ridge_rmse = np.sqrt(mean_squared_error(y_test, ridge_pred))
round(ridge_rmse, 3)

0.088

There is no change to the RMSE when evaluated on the test set with an alpha value of 0.4

### Question 19:
#### Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [29]:
lasso = Lasso(alpha=0.001)
lasso.fit(X_train,y_train)
lasso_predicted_values= lasso.predict(X_test)

lasso_getweight_df =get_weights_df(lasso,X_train,'Lasso_weight')
non_zero_weight= lasso_getweight_df[lasso_getweight_df['Lasso_weight']!=0]

non_zero_weight

Features  Lasso_weight
0      RH_out     -0.049557
1        RH_8     -0.000110
24  Windspeed      0.002912
25       RH_1      0.017880

There are 4 features having non-zero feature weights.

### Question 20:
#### What is the new RMSE with the lasso regression? (in three (3) decimal places)

In [30]:
# Calulate the Root Mean Squared error (RMSE)
rsme_lasso =round(np.sqrt(mean_squared_error(y_test, lasso_predicted_values)), 3)
rsme_lasso

0.094

The new RMSE with lasso regression is 0.094